### PR TITLE
Fold chains of successive goto statements

### PR DESCRIPTION
**Pull request by vchuravy**
Motivated by @Roger-luo YaoCompiler, we implemented [constant propagation](https://github.com/QuantumBFS/YaoCompiler.jl/pull/9) on SSAIR and ended up with long chains of successive `Goto` statements. This is problematic for Yao since it would like to cluster quantum and classical statements and interspersed control flow makes that harder.

### CODE REVIEW

**Review by Keno — APPROVED** [review]
LGTM - I did think this was already implemented somewhere, but maybe it's in a PR that never made it to master.

**Comment by Keno** [comment]
No, I'm right, we do have it: https://github.com/JuliaLang/julia/blob/master/base/compiler/ssair/passes.jl#L1044

File: base/compiler/ssair/ir.jl [hunk 1]
```diff
@@ -546,30 +546,58 @@ mutable struct IncrementalCompact
...
                 if bb_unreachable(domtree, i)
-                    bb_rename[i] = -1
+                    bb_rename[i] = -1 # Dead block
                 else
-                    bb_rename[i] = cur_bb
-                    cur_bb += 1
+                    preds = blocks[i].preds
+                    # Is this block the sole successor of it's sole predecessor?
+                    # If yes we can merge them.
+                    if length(preds) == 1 && length(blocks[first(preds)].succs) == 1
```
Comment (Keno) [inline base/compiler/ssair/ir.jl:554]: This doesn't quite work. You can only do this if the three blocks in question are sitting in statement order or if the block in question only has one statement. You need to check those conditions here. Anything that needs to rearrange instructions or rewrite phi nodes is too complicated for IncrementalCompact.

**Review by Keno — CHANGES_REQUESTED** [review]
Marking as changes requested because of the mentioned issue.

**Comment by Roger-luo** [comment]
I can confirm `cfg_simplify!` works for YaoCompiler, so I'm wondering which should we use in YaoCompiler now?

**Comment by vchuravy** [comment]
Oof, how did I not see `cfg_simplify` before implementing this.

**Pull request closed without merging**